yt/frontends/gadget/io.py

```diff
@@ -77,13 +80,58 @@ def _yield_coordinates(self, data_file, needed_ptype=None):
             yield key, pos
         f.close()
 
+    def _generate_smoothing_length(self, data_files, kdtree):
+        if not self.ds.gen_hsmls:
+            return
+        # TODO: We need to do something like this here but I'm not sure how
+        """
+        if os.path.exists(self.hsml_filename):
+            with open(self.hsml_filename, 'rb') as f:
+                file_hash = struct.unpack('q', f.read(struct.calcsize('q')))[0]
+            if file_hash != self.ds._file_hash:
+                os.remove(self.hsml_filename)
```

Comment:
If you're going to remove a file conditionally, I believe it should be logged as a warning (mentioning why it's done too). Also I think the attribute should be dealt with (either set to None or deleted ?)

Reply:
See my comment below, all of this needs to be changed. This is the way it is handled in the Tipsy frontend.

Reply:
So I *think* that you might be able to replace the logic with:

1. Allowing an HSML file name to be generated on a per-file basis
2. Checking if that filename exists when thinking about reading it
3. Replacing the `file_hash` with an h5py file open.

Reply:
What are we comparing the h5py file open to? Do we need to write something in the hsml file to check? There's some logic in the Tipsy frontend I don't fully understand having to do with packing a struct with "q"

Reply:
So `q` is a `long long` and I would guess that tipsy is storing a hash.  It generates the hash in its own frontend, then stores it, and if it's different, it regenerates.

I think we don't need to do the file hash thing here, unless we already have a file hash, and if we do, then go ahead and make it an `attr` on the root node of the hdf5 file.

Reply:
I think we'd have to do that if we don't want to regenerate the file, tbh. So we could use the UUID for the dataset or something

Reply:
Or, oops, the `_file_hash`